src/analysis/indices/spi.py

```diff
@@ -0,0 +1,226 @@
+
+    @staticmethod
+    def stack_pixels(da: xr.DataArray) -> xr.DataArray:
+        return da.stack(point=('lat', 'lon')).groupby('point')
+
+    def init_distribution(self, distribution: str) -> Enum:
+        if distribution == 'gamma':
+            dist = climate_indices.indices.Distribution.gamma
+        elif distribution == 'pearson':
+            dist = climate_indices.indices.Distribution.pearson
+        else:
+            assert False, f"{distribution} is not a valid distribution fit for SPI"
+
+        self.distribution = distribution
+
+        return dist
+
+    def init_start_year(self, data_start_year: Optional[int] = None) -> int:
+        if data_start_year is None:
+            data_start_year = int(self.ds['time.year'].min().values)
+            print(f"Setting the data_start_year automatically: {data_start_year}")
+
+        assert isinstance(data_start_year, int), f"Expected data_start_year to be an integer"
+        return data_start_year
+
+    def init_calib_year(self, initial_yr: Optional[int],
+                        final_yr: Optional[int]) -> Tuple[int, int]:
+        if initial_yr is None:
+            initial_yr = int(self.ds['time.year'].min().values)
+
+            print(f"Setting the inital timeperiod for calibration manually:\n\
+            inital year: {initial_yr}")
+
+        if final_yr is None:
+            final_yr = int(self.ds['time.year'].max().values)
+
+            print(f"Setting the final timeperiod for calibration manually:\n\
+            final year: {final_yr}")
+
+        assert initial_yr >= int(self.ds['time.year'].min().values), f"intial_year\
+        must be greater than the minimum year of the data\
+        {initial_yr} >= {int(self.ds['time.year'].min().values)}"
+        assert final_yr <= int(self.ds['time.year'].max().values), f"final_year\
+        must be less than the maximum year of the data .\
+        {final_yr} <= {int(self.ds['time.year'].max().values)}"
+
+        return initial_yr, final_yr
+
+    def init_periodicity(self, periodicity: Optional[str]) -> Enum:
+        if periodicity is None:
+            periodicity = 'gamma'
+
+        if periodicity == 'monthly':
+            period = climate_indices.compute.Periodicity.monthly
+        elif periodicity == 'daily':
+            period = climate_indices.compute.Periodicity.daily
+        else:
+            assert False, f"{periodicity} is not a valid periodicity for SPI"
+
+        self.periodicity = periodicity
+
+        return period
+
+    def initialise_params(self,
+                          scale: int,
+                          distribution: str,
+                          data_start_year: Optional[int],
+                          calibration_year_initial: Optional[int],
+                          calibration_year_final: Optional[int],
+                          periodicity: Optional[str]
+                          ) -> Tuple[int, Enum, int, int, int, int, Enum]:
```

Comment:
and this is another place causing the `mypy` errors 

```
src/analysis/indices/spi.py:146: error: Incompatible return value type (got "Tuple[int, Enum, int, int, int, Callable[[Optional[int], Optional[int]], Tuple[int, int]], Enum]", expected "Tuple[int, Enum, int, int, int, int, Enum]")
```

Reply:
this one I don't really understand because from the error message it looks to me like their suggested length of return values is longer than the return values?

Reply:
(Assuming these have been resolved since the tests pass)